### PR TITLE
Translator: translate current page

### PR DESCRIPTION
**Pull request by hius07**
Available via the menu and with a gesture.
Forces book language (if available) as a source language.
Shows main translation only, without source text, alternative translations and definitions.
Just a page to read.
Closes https://github.com/koreader/koreader/issues/10199.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10438)

### CODE REVIEW

File: frontend/ui/translator.lua [hunk 1]
```diff
@@ -540,12 +579,14 @@ function Translator:_showTranslation(text, target_lang, source_lang, from_highli
             table.insert(source, s)
             table.insert(translated, t)
         end
-        table.insert(output, "▣ " .. table.concat(source, " "))
-        text_main = "● " .. table.concat(translated, " ")
+        if not full_page then
+            table.insert(output, "▣ " .. table.concat(source, " "))
+        end
+        text_main = (full_page and "" or "● ") .. table.concat(translated, " ")
```
Comment (poire-z) [inline frontend/ui/translator.lua:585]: (When naming this new parameter "full_page", did you think about translating the current full page - or showing the result translation full page ? I would have named it "no_alt_translation" or "simple_translation" as you're just ignoring all the other translations/definitions. But may be this ambivalent "full_page", because it is ambivalent, is fine.)

File: frontend/ui/translator.lua [hunk 2]
```diff
@@ -653,6 +694,31 @@ function Translator:_showTranslation(text, target_lang, source_lang, from_highli
     UIManager:show(textviewer)
 end
 
+function Translator:onTranslateCurrentPage()
+    local ui = require("apps/reader/readerui").instance
+    if not ui or not ui.document then return end
+    local x0, y0, x1, y1, page
+    if ui.rolling then
```
Comment (poire-z) [inline frontend/ui/translator.lua:701]: It feels all this "reader" business should better be done in your added:
```lua
function ReaderHighlight:onTranslateCurrentPage()
    Translator:onTranslateCurrentPage()
end
```
It is still ReaderHighlight:onTranslateCurrentPage that will always be called, isn't it?

File: frontend/ui/elements/reader_menu_order.lua [hunk 3]
```diff
@@ -202,6 +202,7 @@ local order = {
         "wikipedia_history",
         "wikipedia_settings",
         "----------------------------",
+        "translate_page",
```
Comment (poire-z) [inline frontend/ui/elements/reader_menu_order.lua:205]: This will add a 11th item in that menu (10th for me with Vocabulary builder disabled), so making it 2 pages :/
No real better idea. In the 2nd menu with a page icon? so-so.
May be the dict & wiki lookup history should be moved into their respecting ... settings, or be in the own submenu?

Comment (hius07) [reply frontend/ui/elements/reader_menu_order.lua:205]: 12 (incl. calibre search).
Maybe move dict, wiki and translation settings into Settings submenu?

File: frontend/ui/translator.lua (hunk 2 above)
Comment (hius07) [reply frontend/ui/translator.lua:701]: Sure, it must be done in ReaderHighlight.

**Comment by hius07** [comment]
Closes https://github.com/koreader/koreader/issues/10199.

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 4]
```diff
@@ -466,8 +468,20 @@ function ReaderHighlight:addToMainMenu(menu_items)
             sub_item_table = self:genPanelZoomMenu(),
         }
     end
+
+    -- main menu Search
+    menu_items.search_settings = { -- submenu with Dict, Wiki, Translation settings
+        text = _("Settings"),
+    }
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerhighlight.lua:475]: May be this one should better be in readerdictionary, which is the one that brings into it the most subitems and the first in that search menu? (ReaderHighlight feels a bit unrelated).

File: frontend/dispatcher.lua [hunk 5]
```diff
@@ -340,6 +342,8 @@ local dispatcher_menu_order = {
     "book_description",
     "book_cover",
 
+    "translate_page",
```
Comment (poire-z) [inline frontend/dispatcher.lua:345]: (OK about having it here, can't find a better place in the Reader> actions menu.)

File: frontend/apps/reader/modules/readerhighlight.lua (hunk 4 above)
Comment (hius07) [reply frontend/apps/reader/modules/readerhighlight.lua:475]: What about inconsistency `Dictionary lookup history` vs `Wikipedia history`?

Comment (poire-z) [reply frontend/apps/reader/modules/readerhighlight.lua:475]: Dunno when/where this difference comes from :/ @Frenzie will tell us if/how we should fix that.

Comment (Frenzie) [reply frontend/apps/reader/modules/readerhighlight.lua:475]: I don't know where it came from, but I think "Dictionary history" might be a bit ambiguous?

File: frontend/ui/translator.lua (hunk 1 above)
Comment (poire-z) [reply frontend/ui/translator.lua:577]: I think this snippet would read better if you kept the 2 ways separate, rather than `(full_page and "" or "● ") ..`:
```lua
if full_page then
        text_main = table.concat(translated, " ")
else
        table.insert(output, "▣ " .. table.concat(source, " "))
        text_main = "● " .. table.concat(translated, " ")
end
```

File: frontend/ui/translator.lua [hunk 6]
```diff
@@ -526,6 +557,7 @@ function Translator:_showTranslation(text, target_lang, source_lang, from_highli
     end
     local output = {}
     local text_main = ""
+    local not_full_page = not full_page
```
Comment (poire-z) [inline frontend/ui/translator.lua:560]: This feels ugly and prone to typo :)
And it means "detailed".
So, may be name it everywhere "detailed", and if the default should be true:
`if detailed == nil then detailed = true end`
and pass `false` when you don't want it.

File: frontend/ui/translator.lua [hunk 7]
```diff
@@ -477,31 +505,34 @@ Show translated text in TextViewer, with alternate translations
...
     -- translation service query.
     local Trapper = require("ui/trapper")
     Trapper:wrap(function()
-        self:_showTranslation(text, target_lang, source_lang, from_highlight, page, index)
+        self:_showTranslation(text, target_lang, source_lang, from_highlight, page, index, translate_page)
     end)
 end
 
-function Translator:_showTranslation(text, target_lang, source_lang, from_highlight, page, index)
+function Translator:_showTranslation(text, target_lang, source_lang, from_highlight, page, index, translate_page)
```
Comment (poire-z) [inline frontend/ui/translator.lua:530]: Just some thoughts (not insisting too much, but just in case you'd agree :):
`translate_page` still feels a bit odd.
Translator can be seen as some utility tool, and shouldn't really be that linked to the idea of translating a _book_ page or a fragment of a _book_ content. It could be used to translate a Book description, or some Wikipedia result content.
So, ideally, the params shouldn't indicate what's the source (with some inner logic to decide what's expected as the result), but directly what's expected. So, "detailed" vs "simple".
We do pass from_highlight/page/index, so yes, there is a bit of outer context, but that's just for adding buttons.
(Isn't from_highlight a synonym of translate_fragment, ie. it's not there when you translate a page? Or can we don't have it even with fragments?)

**Comment by poire-z** [comment]
Looks like "[x] Auto detect source language" has no longer the intended effect.
On a French book, highlighting a german sentence, and translate'ing, I get:
`Calling https://translate.googleapis.com/translate_a/single?sl=fr&oe=UTF-8&hl=en&ssel=0&tsel=0&dt=t&dt=at&dt=md&client=gtx&tl=fr&ie=UTF-8&otf=1&q=the-words-in-german`
and I get the title "Translation from French".
Previously (on an older appimage), I got "Translation from German".

**Comment by Frenzie** [comment]
Sounds like Google autodetect got worse? ;-P
